dftimewolf/lib/collectors/gcp_logging.py

```diff
@@ -42,75 +40,121 @@ def __init__(self,
       filter_expression (str): GCP advanced logs filter expression.
+      backoff (bool): Retry queries with an increased delay when API quotas are exceeded.
+      delay (str): Seconds to wait between retreiving results pages to avoid exceeding API quotas
     """
     self._project_name = project_name
     self._filter_expression = filter_expression
+    self._backoff = backoff
+    self._delay = float(delay)
+    self.PublishMessage(f'backoff {self._backoff}')
+    self.PublishMessage(f'delay {self._delay}')
 
   def Process(self) -> None:
     """Copies logs from a cloud project."""
 
-    output_file = tempfile.NamedTemporaryFile(
-        mode='w', delete=False, encoding='utf-8', suffix='.jsonl')
-    output_path = output_file.name
-    self.logger.info(f'Downloading logs to {output_path:s}')
+    output_file, output_path = self.OutputFile()
 
     try:
-      if self._project_name:
-        logging_client = logging.Client(_use_grpc=False,  # type: ignore
-                                        project=self._project_name)
-      else:
-        logging_client = logging.Client(_use_grpc=False)  # type: ignore
-
-      results = logging_client.list_entries(  # type: ignore
-          order_by=logging.DESCENDING,
-          filter_=self._filter_expression,
-          page_size=1000)
+      'Setup a logging client'
+      logging_client = self.SetupLoggingClient()
+      
+      'Get a generator of query results'
+      pages = self.ListPages(logging_client)
 
-      pages = results.pages
-
-      have_results = True
-      while have_results:
-        have_page = False
-        page = []
-        retries = 0
-        while not have_page and retries < MAX_RETRIES:
-          try:
-            page = next(pages)
-            have_page = True
-          except google_api_exceptions.TooManyRequests as exception:
-            retries += 1
-            self.logger.warning(
-                'Hit quota limit requesting GCP logs (retries {0:d} of {1:d}): '
-                '{2:s}'.format(
-                    retries, MAX_RETRIES, str(exception)))
-            time.sleep(4)
-            continue
-          except StopIteration:
-            have_results = False
-            break
-
-        if not have_page and retries >= MAX_RETRIES:
-          self.ModuleError(
-              f'Hit max retries ({MAX_RETRIES:d}) requesting GCP logs',
-              critical=True)
-
-        for entry in page:
-
-          log_dict = entry.to_api_repr()
-          output_file.write(json.dumps(log_dict))
-          output_file.write('\n')
+      'Iterate through query result pages and save json logs to disk'
+      output_path = self.ProcessPages(pages, 2, output_file, output_path)
 
     except google_api_exceptions.NotFound as exception:
       self.ModuleError(
-          f'Error accessing project: {exception!s}', critical=True)
+          'Error accessing project: {0!s}'.format(exception), critical=True)
```

Comment:
Similar to above regarding fstrings, but perhaps I'm missing a reason you changed this line?

Reply:
There was no reason that I recall. I've changed this back - thanks